crates/sui-types/src/programmable_transaction_builder.rs

```diff
@@ -31,29 +37,54 @@ impl ProgrammableTransactionBuilder {
 
     pub fn finish(self) -> ProgrammableTransaction {
         let Self { inputs, commands } = self;
-        let inputs = inputs.into_iter().collect();
+        let inputs = inputs.into_values().collect();
         ProgrammableTransaction { inputs, commands }
     }
 
+    fn pure_bytes(&mut self, bytes: Vec<u8>) -> Argument {
+        let n = self.inputs.len();
+        let (i, _) = self
+            .inputs
+            .insert_full(BuilderArg::Pure(n), CallArg::Pure(bytes));
+        Argument::Input(i as u16)
+    }
+
     pub fn pure<T: Serialize>(&mut self, value: T) -> anyhow::Result<Argument> {
-        Ok(self
-            .input(CallArg::Pure(
-                bcs::to_bytes(&value).context("Searlizing pure argument.")?,
-            ))
-            .unwrap())
+        Ok(self.pure_bytes(bcs::to_bytes(&value).context("Searlizing pure argument.")?))
     }
 
-    pub fn obj(&mut self, obj_arg: ObjectArg) -> Argument {
-        self.input(CallArg::Object(obj_arg)).unwrap()
+    pub fn obj(&mut self, obj_arg: ObjectArg) -> anyhow::Result<Argument> {
+        let id = obj_arg.id();
+        let (i, old_value_opt) = self
+            .inputs
+            .insert_full(BuilderArg::Object(id), CallArg::Object(obj_arg));
+        if let Some(old_value) = old_value_opt {
+            let obj_arg = CallArg::Object(obj_arg);
+            anyhow::ensure!(
+                old_value == obj_arg,
+                "Mismatched Object argument kind for object {i}. \
+                {old_value:?} is not compatible with {obj_arg:?}"
+            )
```

Comment:
I assume I need a follow up PR to handle the special case where `obj_arg` is the same as `old_value` except that the `mutable` field is different? It's legal to have two commands that reference the same shared object but one with mutable reference and the other with immutable reference per @amnn 's [comment](https://github.com/MystenLabs/sui/pull/8737#discussion_r1132300973) right?

Reply:
I could just fix that here... hold on